File: moderate/balanced_smileys.pl
use strict;

sub balanced {
    my ( $s, $c ) = @_;
    while ( $c >= 0 ) {
        return ( $c == 0 ) if ( !( $s =~ /[:()]/ ) );
        return ( balanced( substr( $s, 2 ), $c )
              || balanced( substr( $s, 2 ), $c + 1 ) )
          if ( substr( $s, 0, 2 ) eq ':(' );
        return ( balanced( substr( $s, 2 ), $c )
              || balanced( substr( $s, 2 ), $c - 1 ) )
          if ( substr( $s, 0, 2 ) eq ':)' );
        $c++ if ( substr( $s, 0, 1 ) eq '(' );
        $c-- if ( substr( $s, 0, 1 ) eq ')' );
        $s =~ s/^.[^:()]*//;
    }
    return 0;
}
open( INFILE, "<", $ARGV[0] )
  or die("Cannot open file $ARGV[0] for reading: $!");
while ( my $line = <INFILE> ) {
    printf( "%s\n", balanced( $line, 0 ) ? "YES" : "NO" );
}
close(INFILE);

File: moderate/balanced_smileys.rb
def balanced(s, c)
  while c >= 0
    return c == 0 if s.empty?
    return balanced(s[2..-1], c) || balanced(s[2..-1], c + 1) if s[0..1] == ':('
    return balanced(s[2..-1], c) || balanced(s[2..-1], c - 1) if s[0..1] == ':)'
    c += 1 if s[0] == '('
    c -= 1 if s[0] == ')'
    s = s[1..-1]
  end
  false
end

File.open(ARGV[0]).each_line do |line|
  puts balanced(line.chomp, 0) ? 'YES' : 'NO'
end
